scripts/asr_language_modeling/ngram_lm/train_kenlm.py

```diff
@@ -54,113 +53,164 @@
-        )
-        encoding_level = 'char'
+    tokenizer, encoding_level, is_aggregate_tokenizer = kenlm_utils.setup_tokenizer(args.tokenizer_model_file)
 
-    """ DATASET SETUP """
-    logging.info(f"Encoding the train file '{args.train_file}' ...")
-    dataset = kenlm_utils.read_train_file(args.train_file, lowercase=args.do_lowercase)
-    encoded_train_file = f"{args.kenlm_model_file}.tmp.txt"
     if encoding_level == "subword":
-        kenlm_utils.tokenize_text(
-            dataset,
-            model.tokenizer,
-            path=encoded_train_file,
-            chunk_size=CHUNK_SIZE,
-            buffer_size=CHUNK_BUFFER_SIZE,
-            token_offset=DEFAULT_TOKEN_OFFSET,
-        )
-        # --discount_fallback is needed for training KenLM for BPE-based models
-        discount_arg = "--discount_fallback"
+        discount_arg = "--discount_fallback"  # --discount_fallback is needed for training KenLM for BPE-based models
     else:
-        with open(encoded_train_file, 'w', encoding='utf-8') as f:
-            for line in dataset:
-                f.write(f"{line}\n")
-
         discount_arg = ""
 
-    del model
-
     arpa_file = f"{args.kenlm_model_file}.tmp.arpa"
     """ LMPLZ ARGUMENT SETUP """
     kenlm_args = [
         os.path.join(args.kenlm_bin_path, 'lmplz'),
         "-o",
-        f"{args.ngram_length}",
-        "--text",
-        encoded_train_file,
+        str(args.ngram_length),
         "--arpa",
         arpa_file,
         discount_arg,
-    ]
+        "--prune",
+    ] + args.ngram_prun
+
+    if args.cache_path:
+        if not os.path.exists(args.cache_path):
+            os.makedirs(args.cache_path, exist_ok=True)
+
+        """ DATASET SETUP """
+        encoded_train_files = []
+        for file_num, train_file in enumerate(train_path):
+            logging.info(f"Encoding the train file '{train_file}' number {file_num+1} out of {len(train_path)} ...")
+
+            cached_files = glob(os.path.join(args.cache_path, os.path.split(train_file)[1]) + "*")
+            encoded_train_file = os.path.join(args.cache_path, os.path.split(train_file)[1] + f"_{file_num}.tmp.txt")
+            if cached_files:
+                if cached_files[0] != encoded_train_file:
+                    os.rename(cached_files[0], encoded_train_file)
+                    logging.info("Rename", cached_files[0], "to", encoded_train_file)
+
+            encoded_train_files.append(encoded_train_file)
+
+        kenlm_utils.iter_files(
+            encoded_train_files,
+            train_path,
+            tokenizer,
+            encoding_level,
+            is_aggregate_tokenizer,
+            args.do_lowercase,
+            args.rm_punctuation,
+            args.separate_punctuation,
+            args.verbose,
+        )
 
-    ret = subprocess.run(kenlm_args, capture_output=False, text=True, stdout=sys.stdout, stderr=sys.stderr)
-    if ret.returncode != 0:
+        first_process_args = ["cat"] + encoded_train_files
```

Comment:
Make the output of subprocesses print onto sys out and err - otherwise hard to figure out what went wrong

Reply:
stdout goes to KenLm, stderr is fixed